### ISSUE TITLE
Question: how come function references on the global scope are not closed

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint-scope#57]
https://github.com/eslint/eslint-scope/blob/dbddf14d5771b21b5da704213e4508c660ca1c64/tests/references.js#L211-L263

I'm working on trying to understand this codebase better.
But I can't understand why it works this way:

```js
function a() {}
a();
```

To me it seems clear that the reference that's created as part of the call should resolve directly to the function declaration, but it's not - it remains as an unresolved reference that's added to the global scope's `through` list.

This would cause the `no-unused-vars` eslint rule to error on the function declaration, were it not for [this code](https://github.com/eslint/eslint/blob/015edf6467e33c67b904db037a674d71957a6865/lib/linter/linter.js#L172-L194) in the linter, which just so happens to pick up this unresolved reference, and forcefully resolves it to the function declaration (seemingly by chance?).

Is anyone able to explain why this works this way?

Or rather - how come when closing the global scope it attempt to resolve all `through` against the variables defined in the global scope?

**Comment by mysticatea** [comment]
Thank you for your question.

....but, I'm not sure why it's so. I guess that it's as-is from the original package `escope`.

Maybe... it couldn't assume declarations in global scope make variables, because browsers may ignore declarations.

```js
var top = function() {}
top() // ERROR: 'top' is not a function
```

> `data:text/html,<script>var top = function(){}; top();</script>`

**Comment by bradzacher** [comment]
Interesting.
Is this something that should be fixed?

I believe at least in `sourceType: module` mode, this shouldn't work this way, right?

**Comment by mysticatea** [comment]
Yes, `data:text/html,<script type="module">var top = function(){}; top();</script>` works fine.

Also, new `knownGlobals: string[]` option may be good to make variables and resolve references.

**Comment by bradzacher** [comment]
is it only `var` declarations that shouldn't be closed?

Doing some testing it looks like every single other case works as expected and throws this error: `Identifier 'top' has already been declared`

It's only `var top` that is ignored and throws `top is not a function`

**Comment by nzakas** [comment]
@mysticatea what's the verdict here? Is there something to fix or no?

**Comment by nzakas** [comment]
@mysticatea is this something you want to address?
